Wallet Closure (#327)
keep-network/tbtc-v2: README.adoc

`Wallets.registerNewWallet`, which sets the new wallet as the active wallet.
Going forward, it will receive deposits.

==== Wallet Closure

Wallets can close in a few ways:

> * The operators notify the chain that the wallet is failing a heartbeat: `Wallets.notifyWalletHeartbeatFailed`

Reviewer: ```suggestion
* The operators notify the chain that the wallet is failing a heartbeat: `Bridge. __ecdsaWalletHeartbeatFailedCallback` (called by `ecdsa` `WalletRegistry.notifyOperatorInactivity`).
```

> * Someone notifies the chain that the wallet timed out while filling a redemption: `Wallets.notifyWalletTimedOutRedemption`

Reviewer: ```suggestion
* Someone notifies the chain that the wallet timed out while filling a redemption: `Bridge.notifyRedemptionTimeout`
```

Author: [7f1c41e](https://github.com/keep-network/tbtc-v2/pull/327/commits/7f1c41ede14ed9f42daf015a83de41c117dba994)

> * Someone notifies the chain that the wallet is too old: `Wallets.notifyCloseableWallet`
> * Someone notifies the chain that the wallet has too few BTC remaining: `Wallets.notifyCloseableWallet`

Reviewer: ```suggestion
* Someone notifies the chain that the wallet is too old: `Bridge.notifyCloseableWallet`
* Someone notifies the chain that the wallet has too few BTC remaining: `Bridge.notifyCloseableWallet`
```

Each of the above routes into `Wallets.moveFunds`. If the wallet has no main
UTXO, then we can start closing it immediately via
`Wallets.beginWalletClosing`. Otherwise, we change it's state to
`WalletState.MovingFunds`, decrease the live wallet count, and give the wallet
a deadline to move its funds to other wallet(s). If this was the active wallet,

> then we currently have no active wallet.

Reviewer: The active wallet can not be closed by `notifyCloseableWallet`:
https://github.com/keep-network/tbtc-v2/blob/f1076f450a05c6b87c91e2f49a51fb43799f5439/solidity/contracts/bridge/Wallets.sol#L343-L346

We require another wallet to be created to replace it.

Author: well hot dang!

Author: The other routes don't have that protection though:

https://github.com/keep-network/tbtc-v2/blob/main/solidity/contracts/bridge/Wallets.sol#L399

If we get into `movingFunds` via `Bridge.notifyRedemptionTimeout` or `WalletRegistry.notifyOperatorInactivity` and the wallet is the active wallet, then we're left with no active wallet. I'll amend the last two bullet points!

Author: [b08e80a](https://github.com/keep-network/tbtc-v2/pull/327/commits/b08e80a38d1f2dd31e897fd9f963f7e1902fd270)

Reviewer: Yeah, it's only for `notifyCloseableWallet`. And I think it makes sense - if the wallet is in a good shape (no heartbeat failure), we can still rely on it. Rotating active wallets just because they are old enough and without giving them a chance to accumulate enough wealth could lead to too many move funds operations between wallets.

After `movingFundsTimeout` goes by, anyone can call

> `MovingFunds.notifyMovingFundsTimeout` which pipes into

Reviewer: ```suggestion
`Bridge.notifyMovingFundsTimeout` which pipes into
```

Author: [7f1c41e](https://github.com/keep-network/tbtc-v2/pull/327/commits/7f1c41ede14ed9f42daf015a83de41c117dba994)

`Wallets.terminateWallet` followed by `ecdsaWalletRegistry.seize`.
`terminateWallet`, in this case passes through to
`ecdsaWalletRegistry.closeWallet` after changing the state to
`WalletState.Terminated`.

Furthermore, the `ecdsaWalletRegistry.seize` call is punishing the operators by
`movedFundsSweepTimeoutSlashingAmount` and rewarding the notifier with a reward
multiplier of `movedFundsSweepTimeoutNotifierRewardMultiplier` (the notifier
gets a percentaged of the slashed stake).

In order to avoid this, the wallet has to commit to which wallets they'll send

> the funds to (`MovingFunds.submitMovingFundsCommitment`), then actually send

Reviewer: ```suggestion
the funds to (`Bridge.submitMovingFundsCommitment`), then actually send
```

Author: Making the change!

If `MovingFunds.submitMovingFundsCommitment` is also `external` why do we have the layer of indirection?

Reviewer: `MovingFunds` is a library, `Bridge` is a contract. `MovingFunds` does not have its own storage, it's using `Bridge` storage (`BridgeState`) and we can not call `MovingFunds` directly. The fact it's `external` is just about that it's called via DELEGATECALL and the compiler is not inlining the code of the called function into `Bridge` contract so that we can save on the contract bytecode size (24 KB limit on EVM).

the funds to those wallets, and then prove that they did it

> (`MovingFunds.submitMovingFundsProof`) before the time runs out.

Reviewer: ```suggestion
(`Bridge.submitMovingFundsProof`) before the time runs out.
```

Author: [7f1c41e](https://github.com/keep-network/tbtc-v2/pull/327/commits/7f1c41ede14ed9f42daf015a83de41c117dba994)

The commitment involves submitting a list of wallet public keys that:

* Aren't the source wallet 
* Are in ascending order
* Are Live

We store the hash of the list in `movingFundsTargetWalletsCommitmentHash`.
Off-chain we come to consensus by picking the Live wallets whose public key
hashes are the closest to the source wallet's public key hash in terms of clock
distance (modulus distance). This makes public key hashes in the middle of the
range no more likely to be picked than ones near the ends. We pick a number of
wallets equal to `min(liveWalletsCount, ceil(walletBtcBalance /
walletMaxBtcTransfer))`, where `walletMaxBtcTransfer` is governable.

> In `submitMovingFundsProof`, we prove that the bitcoin transaction happened and

Reviewer: ```suggestion
In `submitMovingFundsProof`, we prove that the Bitcoin transaction happened and
```

Just to keep it aligned with other occurrences.

Author: 🙇

Author: [6461fa3](https://github.com/keep-network/tbtc-v2/pull/327/commits/6461fa3de6bcf9f48a49ffbda8fe3983cc1ff9dd)

`Bridge.__ecdsaWalletCreatedCallback`, which sets the new wallet as the active
wallet. Going forward, it will receive deposits.

==== Wallet Closure

Wallets can close in a few ways:

* The operators notify the chain that the wallet is failing a heartbeat:
`Bridge. __ecdsaWalletHeartbeatFailedCallback` (called by `ecdsa`

> `WalletRegistry.notifyOperatorInactivity`)`

Reviewer: ```suggestion
`WalletRegistry.notifyOperatorInactivity`)
```
